linkcheck/better_exchook2.py

```diff
@@ -40,201 +40,267 @@
-		import traceback
-		for l in traceback.format_exc().split("\n"): output("   " + l, out=out)
-		output("simple traceback:", out=out)
-		traceback.print_tb(tb, None, out)
-
-	import types
-	def _some_str(value):
-		try: return str(value)
-		except: return '<unprintable %s object>' % type(value).__name__
-	def _format_final_exc_line(etype, value):
-		valuestr = _some_str(value)
-		if value is None or not valuestr:
-			line = "%s" % etype
-		else:
-			line = "%s: %s" % (etype, valuestr)
-		return line
-	if (isinstance(etype, BaseException) or
-		(hasattr(types, "InstanceType") and isinstance(etype, types.InstanceType)) or
-		etype is None or type(etype) is str):
-		output(_format_final_exc_line(etype, value), out=out)
-	else:
-		output(_format_final_exc_line(etype.__name__, value), out=out)
+    output('Traceback (most recent call last):', out=out)
+    allLocals, allGlobals = {}, {}
+    try:
+        import linecache
+
+        limit = None
+        if hasattr(sys, 'tracebacklimit'):
+            limit = sys.tracebacklimit
+        n = 0
+        _tb = tb
+
+        def _resolveIdentifier(namespace, id):
+            obj = namespace[id[0]]
+            for part in id[1:]:
+                obj = getattr(obj, part)
+            return obj
+
+        def _trySet(old, prefix, func):
+            if old is not None:
+                return old
+            try:
+                return prefix + func()
+            except KeyError:
+                return old
+            except Exception as e:
+                return prefix + "!" + e.__class__.__name__ + ": " + str(e)
+
+        while _tb is not None and (limit is None or n < limit):
+            f = _tb.tb_frame
+            allLocals.update(f.f_locals)
+            allGlobals.update(f.f_globals)
+            lineno = _tb.tb_lineno
+            co = f.f_code
+            filename = co.co_filename
+            name = co.co_name
+            output('  File "%s", line %d, in %s' % (filename, lineno, name), out=out)
+            if not os.path.isfile(filename):
+                altfn = fallback_findfile(filename)
+                if altfn:
+                    output(
+                        "    -- couldn't find file, trying this instead: " + altfn,
+                        out=out,
+                    )
+                    filename = altfn
+            linecache.checkcache(filename)
+            line = linecache.getline(filename, lineno, f.f_globals)
+            if line:
+                line = line.strip()
+                output('    line: ' + line, out=out)
+                output('    locals:', out=out)
+                alreadyPrintedLocals = set()
+                for tokenstr in grep_full_py_identifiers(parse_py_statement(line)):
+                    splittedtoken = tuple(tokenstr.split("."))
+                    for token in map(
+                        lambda i: splittedtoken[0:i], range(1, len(splittedtoken) + 1)
+                    ):
```

Comment:
I don't know if we want to touch this code, but
```suggestion
                    for token in (
                        splittedtoken[0:i] for i in range(1, len(splittedtoken) + 1)
                    ):
```